### PR TITLE
`Programming exercises`: Improve details page layout

### PR DESCRIPTION
**Pull request by florian-glombik**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
Unify the layout of the programming exercise details- and edit-page by displaying information in the details page in the same sections as on the edit page.
Reducing the scope of some sections by introducing the new section `Build Details`, which is displayed directly after the `General` section, making the repository, checkout directory and `Repository Diff` more prominent on the details page.

### Description
* Moved the `short name`, `auxiliary repositories`, and `recreate build plans` to a new section `Build Details` that is added as second section
* Moved the selection of categories up in the edit mode (within the `General` section) to be aligned with the details page

### Steps for Testing
Prerequisites:
- 1 Instructor

1. View the details page of a programming exercise
2. See that a new section `Build Status` is introduced
3. Go to the create/edit page of a programming exercise 
4. See that the section `Build Status` is reflected in the create/edit page
5. Verify that creating / editing / importing exercises still works

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots

| Change                                                 | Before                                                                                    | After                                                                                     |
|--------------------------------------------------------|-------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------|
| Added Build Details Section in Details View            | image | image |
| Added Build Details Section in Create/Edit/Import View | image | image |

## Summary by CodeRabbit

- **New Features**
	- Introduced a method for displaying detailed build configuration information for programming exercises.
	- Enhanced interaction by integrating a build details component, providing real-time validation feedback during the update process.
	- Improved organization and clarity of the programming exercise detail presentation, enhancing user experience.
	- Added a user-friendly interface for managing build configurations and auxiliary repositories.

- **Bug Fixes**
	- Resolved issues with form validity tracking across multiple sections of the programming exercise update interface.

- **Chores**
	- Removed auxiliary repository management components to streamline the user interface.
	- Added a new build details component to improve functionality and user insights.

### CODE REVIEW

File: src/main/webapp/app/exercises/programming/manage/update/update-components/programming-exercise-information.component.html [hunk 1]
```diff
@@ -45,6 +45,23 @@
             }
         }
     </div>
+    @if (!programmingExerciseCreationConfig.isExamMode) {
```
Comment (florian-glombik) [inline src/main/webapp/app/exercises/programming/manage/update/update-components/programming-exercise-information.component.html:48]: Only moved up in the file to be more consistent between the edit and details view

File: src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts [hunk 2]
```diff
@@ -270,6 +270,7 @@ export class ProgrammingExerciseDetailComponent implements OnInit, OnDestroy {
         exercise.buildConfig = this.programmingExerciseBuildConfig;
         return [
             this.getExerciseDetailsGeneralSection(exercise),
+            this.getExerciseDetailsBuildDetailsSection(exercise),
```
Comment (florian-glombik) [inline src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts:273]: Added the new build details section here

File: src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts [hunk 3]
```diff
@@ -310,54 +311,10 @@ export class ProgrammingExerciseDetailComponent implements OnInit, OnDestroy {
         };
     }
 
-    getExerciseDetailsModeSection(exercise: ProgrammingExercise): DetailOverviewSection {
+    getExerciseDetailsBuildDetailsSection(exercise: ProgrammingExercise): DetailOverviewSection {
         return {
-            headline: 'artemisApp.programmingExercise.wizardMode.detailedSteps.difficultyStepTitle',
+            headline: 'artemisApp.programmingExercise.wizardMode.detailedSteps.buildDetailsStepTitle',
```
Comment (florian-glombik) [inline src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts:316]: Moved some of the details from the language section to the new `Build Details` section - the git diff makes that change look more complicated than it is 😅

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts [hunk 4]
```diff
@@ -496,6 +499,10 @@ export class ProgrammingExerciseUpdateComponent implements AfterViewInit, OnDest
                 title: 'artemisApp.programmingExercise.wizardMode.detailedSteps.generalInfoStepTitle',
                 valid: this.exerciseInfoComponent?.formValid ?? false,
             },
+            {
```
Comment (florian-glombik) [inline src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts:502]: Needed for the validation bar
image

File: src/main/webapp/i18n/de/programmingExercise.json [hunk 5]
```diff
@@ -63,15 +63,18 @@
...
                     "languageStepTitle": "Sprache",
-                    "languageStepMessage": "Wähle die gewünschte Programmiersprache aus und konfiguriere zusätzliche Funktionen bezüglich der Build Umgebung",
+                    "languageStepMessage": "Wähle die gewünschte Programmiersprache aus und konfiguriere zusätzliche Funktionen bezüglich der Build Umgebung.",
                     "problemStepTitle": "Aufgabe",
-                    "problemStepMessage": "Erstelle eine detaillierte Aufgabenstellung für die Übung mit klaren Schritten und Erwartungen an die Studierenden",
+                    "problemStepMessage": "Erstelle eine detaillierte Aufgabenstellung für die Übung mit klaren Schritten und Erwartungen an die Studierenden.",
                     "gradingStepTitle": "Bewertung",
-                    "gradingStepMessage": "Definiere die Bewertungskriterien für diese Aufgabe und wie sich das Ergebnis auf die Gesamtnote des Kurses auswirkt"
+                    "gradingStepMessage": "Definiere die Bewertungskriterien für diese Aufgabe und wie sich das Ergebnis auf die Gesamtnote des Kurses auswirkt."
```
Comment (florian-glombik) [inline src/main/webapp/i18n/de/programmingExercise.json:77]: Only added a dot at the end of the sentence

**Review by Jan-Thurner — DISMISSED** [review]
Tested on TS2. The different functionalities still work after rearranging them. The UI when editing an exercise looks more consistent with the Exercise management view.

**Review by sarpsahinalp — DISMISSED** [review]
Tested on ts2. Works as expected!

**Review by JohannesWt — APPROVED** [review]
Thanks for the changes - code looks good now 👍

**Review by az108 — DISMISSED** [review]
Code LGTM after changes

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
Everything seems to work correctly on TS2, ~~but I noticed some errors in the console though~~:
<details>

<image src="https://github.com/user-attachments/assets/d9c28e4c-e80b-4e84-8b77-d8f5f378752e">

</details>


Edit: seems to be only related to one specific exercise

**Review by SimonEntholzer — APPROVED** [review]

**Review by Jan-Thurner — APPROVED** [review]

**Review by rabeatwork — CHANGES_REQUESTED** [review]
Tiny Feedback

File: src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts [hunk 6]
```diff
@@ -747,7 +756,7 @@ export class ProgrammingExerciseDetailComponent implements OnInit, OnDestroy {
     }
 
     /**
-     * Checks if the build configuration is available and sets the windfile if it is, helpful for reliably displaying
+     * Checks if the build configuration is available and sets the windFile if it is, helpful for reliably displaying
```
[the remark this answers is not in the document]
Comment (florian-glombik) [reply src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts:759]: I looked it up, in the respective thesis it is referred to as "Windfile"

**Review by SimonEntholzer — DISMISSED** [review]
reapprove

**Review by Jan-Thurner — DISMISSED** [review]

**Review by JohannesWt — APPROVED** [review]
Reapprove Code 👍

**Review by az108 — APPROVED** [review]
reapprove after small changes 👍 LGTM

**Review by Jan-Thurner — APPROVED** [review]
reapprove

**Review by JohannesWt — APPROVED** [review]
Reapprove code after fixing client test

**Review by rabeatwork — APPROVED** [review]

**Review by krusche — CHANGES_REQUESTED** [review]
I think the changes are not well thought out due to the following reasons

1. The language section includes many build details, in particular docker image and build script. User might be confused why those can be found in the language section. However, they heavily depend on the chosen programming language, therefore it makes sense to first choose the language and then edit those settings.
 
image

2. It was kind of nice to have 5 rather distinct sections like in other exercises. Adding a 6th one somehow breaks this and might lead to issues when looking at the page on mobile devices (did you test this?)

3. We have carefully chosen all section headlines to consist of one word, now we have one with two words "Build Details".

I am open to improve the programming exercise sections, but it must really make sense and should **not** neglect previous discussion and design ideas!


I suggest we discuss again how to split all settings into 5 sections in a way that editing becomes easy.

This change should also be aligned with the new simple mode that I would like to see for beginners that basically leaves out all the more advanced options.

Just to repeat this from a previous Slack discussion, the simple mode could leave out the following in my opinion:

- Channel Name
- Short Name (just make sure it's unique)
- Participation (individual would be default)
- Allow Offline IDE und allow Online Editor
- Mode could then potentially be combined with General in simple mode and only contains the Difficulty
- Project Type and With exemplary dependency
- Enable Static Code Analysis
- Sequential Test Runs
- Customize Build Script
- Submission Policy
- Timeline simplified: not Start Date, no Example Solution Publication Date, no Run Tests after Due Date
- all options under assessment
- Presentation
- Plagiarism Control
- Competencies

**Comment by florian-glombik** [comment]
[Thank you for the detailed feedback!](https://github.com/ls1intum/Artemis/pull/9208#pullrequestreview-2274137012)

1. The name of the new section does not appear to be well-chosen, I agree that it is confusing to have a section `Build Details` and mainly configuring the build in the `Language` section. As we are actually editing the repositories here _(defining the repository names via the short name and optionally adding auxiliary repositories)_, `Repositories` might be a better name for the section and prevent confusion?<br><br>
   The described problem of the interrelation between the sections `Language` and `Repositories` is something we should discuss for the auxiliary repositories, as the chosen language might have an impact on the configuration of the checkout directories (which is the case for the previous and the new solution), for this advanced case it would be more intuitive to define the language and package name first. In other cases, where no checkout directories are manually defined, I am not aware of differences for the user in the configuration based on the language selection.
2. I was not aware that having 5 sections was an active design decision. While the number of sections would be aligned with text-, modeling- and file-upload-exercises I am not sure if unifying the number of sections prevails the clarity in the programming exercise overview, as currently the `Language` section can be quite long in the exercise details view, and we might be able to allow for a quicker and more intuitive access to chosen configurations by deriving another section, which is the goal of this PR _(together with displaying the `Review Changes`/`Template-Solution-Diff` feature at a more prominent place in the details section)_.<br>
   <br>
   I did not test the responsiveness beforehand, as I relied on the extended `form-status-bar` to handle width changes properly. I did now test it with the following result:
   * With the 6th section `Build Status` (12 chars, such as Repositories)
      * horizontal scrollbar visible for the width 767px - 790px, but section titles still clickable without scroll (2 Tablets: [Acer Picasso, Apple iPad Mini](https://screensiz.es/tablet), 6 phones: [LG Nexus 4, LG Nexus 4 AR, BlackBerry Z10, Nokia Lumia 1020, Nokia Lumia 1020, Nokia Lumia 920, Nokia Lumia 925](https://screensiz.es/phone), no issue for half screen on 14-inch device) 
      * horizontal scrollbar from a width smaller than 540px (phones)

     This is how it looks like 1px before the breakpoint is applied (767px):
     image
   * With the existing 5 sections
      * horizontal scrollbar from a width smaller than 421px (phones)
   
   If the horizontal scrollbar for 767px - 790px widths is an issue, we might be able to resolve this by improving the responsiveness of the `form-status-bar` or adjusting the breakpoint in this case. For phones, I would argue that it doesn't make a reasonable difference.
3. Ramona mentioned that the section titles have been intentionally chosen to be only one word. I misunderstood that this is the case to ensure that the section titles are short and easy to memorize. I was not aware that a single word for the section title is a fixed requirement.   

Thanks for pointing out the previous thoughts that would have been violated! I am sorry about neglecting previous ideas with this PR, which was not my intention. I was not aware of #2 and only touched #3 in a discussion with Ramona, after which I thought `Build Details` would be a reasonable section title.

### Follow-up Questions

For my understanding the reduced configuration options of the `Simple Mode` should be reflected in the details view, which would result in the `Language` section being reduced anyways (`Sequential Test Runs`, `Build Script`, and `Docker Image` would not be displayed in the details view). Do we have the same understanding?<br>

If we decide to limit ourselves to 5 sections, I do think that the sections `Mode`, `Problem`, and `Grading` are well-defined and blend in nicely with the other exercise types. This would leave the open question if we want to move the definition of the `Short Name` and `Auxiliary Repositories` to the `Language` section, due to the dependence on the `Programming Language` and `Package Name` selection.
* pro
   * might be more logical when editing
   * more consistent with the details view, where the repositories are already listed under the language section
* contra
   * makes the configuration of the `Language` section longer in advanced mode

I would be in favor of moving the short name and auxiliary repository selection to the language section, as I would rate the pro bullet-points as more important. However, I have no strong opinion on this topic.

**Pull request closed without merging**